compiler/dyno/test/resolution/testReturnTypes.cpp

```diff
@@ -0,0 +1,356 @@
+  });
+}
+
+static void test3() {
+  // test returning a param from a param-returning function
+  testProgram({
+      lit("1")
+  }, [](auto& qt) {
+    assert(qt.type() && qt.type()->isIntType());
+    assert(qt.kind() == QualifiedType::PARAM && qt.param());
+    assert(qt.param()->toIntParam()->value() == 1);
+  }, QualifiedType::PARAM);
+}
+
+static void test4() {
+  // test returning a param from an ambigous param-returning function
+  // non-ambigous tests are in testParamIf.
+  testProgram({
+      lit("1"),
+      lit("2")
+  }, [](auto& qt) {
+    assert(qt.isUnknown());
+  }, QualifiedType::PARAM);
+}
+
+static void test5() {
+  // test allowed coercions
+  testProgram({
+      decl("const", "int(32)"),
+      decl("const", "int(16)")
+  }, [](auto& qt) {
+    assert(qt.kind() == QualifiedType::VAR);
+    assert(qt.type() && qt.type()->isIntType());
+    assert(qt.type()->toIntType()->bitwidth() == 32);
+  });
+}
+
+static void test6() {
+  // test disallowed coercions
+  testProgram({
+      decl("const", "int(32)"),
+      decl("const", "string")
+  }, [](auto& qt) {
+    assert(qt.isUnknown());
+  });
+}
+
+static void test7() {
+  // test param decaying to var
+  testProgram({
+      decl("const", "int(32)"),
+      lit("1")
+  }, [](auto& qt) {
+    assert(qt.kind() == QualifiedType::VAR);
+    assert(qt.type() && qt.type()->isIntType());
+  });
+}
+
+static void test8() {
+  // test param decaying to var, but for strings
+  testProgram({
+      decl("const", "string"),
+      lit("\"hello\"")
+  }, [](auto& qt) {
+    assert(qt.kind() == QualifiedType::VAR);
+    assert(qt.type() && qt.type()->isStringType());
+  });
+}
+
+static void test9() {
+  // test multiple branches (not just two)
+  testProgram({
+      decl("const", "int(16)"),
+      lit("1"),
+      decl("const", "int"),
+      decl("const", "int(32)"),
+  }, [](auto& qt) {
+    assert(qt.kind() == QualifiedType::VAR);
+    assert(qt.type() && qt.type()->isIntType());
+    assert(qt.type()->toIntType()->bitwidth() == 64);
+  });
+}
+
+static void test10() {
+  // returning types from compile-time unknown function
+  testProgram({
+      type("int"),
+      type("int"),
+  }, [](auto& qt) {
+    assert(qt.kind() == QualifiedType::TYPE);
+    assert(qt.type() && qt.type()->isIntType());
+    assert(qt.type()->toIntType()->bitwidth() == 64);
+  }, QualifiedType::TYPE);
+  testProgram({
+      type("int"),
+      type("bool"),
+  }, [](auto& qt) {
+    assert(qt.isUnknown());
+  }, QualifiedType::TYPE);
+}
+
+static void test11() {
+  // test mixing types and values
+  testProgram({
+      type("int"),
+      lit("1"),
+  }, [](auto& qt) {
+    assert(qt.isUnknown());
+  }, QualifiedType::TYPE);
+  testProgram({
+      type("int"),
+      lit("1"),
+  }, [](auto& qt) {
+    assert(qt.isUnknown());
+  });
+}
+
+static void test12() {
+  // test returning non-param from param-intent procedure
+  testProgram({
+      lit("1"),
+      lit("2"),
```

Comment:
from the comment above, shouldn't at least one of these be `decl` ? As it is, aren't we returning params on all the branches?

Reply:
Since the two params have different values, they are made to decay into values (the common type is `const int`). However, this test can be written more clearly by just using two decls (or even a single decl).

Reply:
Ah. Or update the comment `   // test returning non-param from param-intent procedure ` to clarify.